apps/endatix-hub/components/survey-creator.tsx

```diff
@@ -0,0 +1,171 @@
+  const [isSaving, setIsSaving] = useState(false);
+  const router = useRouter();
+  const [creator, setCreator] = useState<SurveyCreator | null>(null);
+  const [isEditingName, setIsEditingName] = useState(false);
+  const [name, setName] = useState(formName);
+  const inputRef = useRef<HTMLInputElement | null>(null);
+  const [originalName, setOriginalName] = useState(formName);
+
+  useEffect(() => {
+    if (!creator) {
+        const newCreator = new SurveyCreator(options || defaultCreatorOptions);
+        newCreator.JSON = formJson;
+        newCreator.saveSurveyFunc = (no: number, callback: (num: number, status: boolean) => void) => {
+        console.log(JSON.stringify(newCreator?.JSON));
+        callback(no, true);
+        };
+        setCreator(newCreator);
+    }
+
+  }, [creator, formJson, options]);
+
+  useEffect(() => {
+    if (isEditingName) {
+      document.addEventListener('mousedown', handleClickOutside);
+    } else {
+      document.removeEventListener('mousedown', handleClickOutside);
+    }
+    return () => {
+      document.removeEventListener('mousedown', handleClickOutside); // Clean up event listener
+    };
+  }, [isEditingName]);
+
+  const handleSaveAndGoBack = async () => {
+    try {
+      setIsSaving(true);
+
+      const updatedFormJson = creator?.JSON;
+
+      const result = await updateFormDefinitionJsonAction(formId, updatedFormJson);
+      if (!result.success) {
+        throw new Error(result.error);
+      }
+
+      router.push('/forms');
+    } catch (error) {
+      console.error('Failed to save form', error);
+    } finally {
+      setIsSaving(false);
+    }
+  };
+
+  const handleNameSave = async () => {
+    if (name !== originalName) {
+      startTransition(async () => {
+        await updateFormNameAction(formId, name);
+
+        setOriginalName(name);
+        setName(name);
+      });
+    }
+    setIsEditingName(false);
+  };
+
+  const handleKeyDown = (e: React.KeyboardEvent) => {
+    if (e.key === 'Enter') {
+      handleNameSave();
+    } else if (e.key === 'Escape') {
+      setName(originalName);
+      setIsEditingName(false);
+    }
+  };
+
+  const handleClickOutside = (e: MouseEvent) => {
+    if (inputRef.current && !inputRef.current.contains(e.target as Node)) {
+      console.log("Clicked outside, waiting to save name...");
+      setTimeout(() => {
+        handleNameSave();
+      }, 0);
+    }
+  };
+
+  const toggleEnabled = async (enabled : boolean) => {
+    startTransition(async () => {
+        await updateFormStatusAction(formId, enabled);
+    })
+  }
+
+  return (
+    <div>
+        <div className="flex items-center mb-4 w-full">
+        <button
+            onClick={handleSaveAndGoBack}
+            className="mr-4 text-3xl flex items-center"
+            disabled={isSaving}
+            style={{ border: "none", background: "transparent" }}
+        >
+            ←
+        </button>
+
+        {isEditingName ? (
```

Comment:
For future feature development, I would recommend using smaller React components to be closer to the Single Responsibility on UI level. E.g. define `TextFieldComponent`, which handles the Editing and Read states including success/error, state mutation and etc. If we add 4-5 more fields this will get to 500+ lines of code, so we need to bear this in mind for improvement.

Just a mark that we should think about these things as we grow the Endatix Hub